release-23.2: changefeedccl: add kafka sink v2

asg0451 opened the pull request:
Backport:
  * 1/1 commits from "changefeedccl: fix sink leaks in a few places" (#126512)
  * 1/1 commits from "changefeedccl: implement new Kafka sink to sidestep reordering issue" (#126213)
  * 1/1 commits from "changefeedccl: fix data race in kafka sink v2 partitioner" (#127585)
  * 1/1 commits from "changefeedccl: fix kafka sink v2 flush frequency" (#127848)

Please see individual PRs for details.

Release justification: The new sink resolves ordering violation issues present in the v1 sink. I plan to backport it to 24.2, 24.1, and 23.2.

/cc @cockroachdb/release

asg0451 committed:
changefeedccl: implement new Kafka sink to sidestep reordering issue

Add a new Kafka sink utilizing the franz-go
library and our batchingSink wrapper that does not
suffer from the row reordering issue that the
sarama one does. It is behind a cluster setting
(`changefeed.new_kafka_sink_enabled`) which is
disabled by default. We intend to backport this
change. To that end, we use the previous version
of franz-go (v1.16.1) to not require bumping
important dependencies.

Epic: CRDB-38892
Resolves: #122372

Release note (enterprise change): Add a new Kafka
sink utilizing the franz-go library and our own
batching_sink behind a cluster setting
(`changefeed.new_kafka_sink_enabled`, disabled by
default).

asg0451 committed:
changefeedccl: fix sink leaks in a few places

There are some places where we neglect to close
sinks, which lead to goroutine leaks and thus test
failures in the new Kafka sink PR.

Epic: CRDB-38892

Release note (bug fix): Fixed small memory leaks
that occur during changefeed creation.

asg0451 committed:
changefeedccl: fix data race in kafka sink v2 partitioner

The partitioner passed to kgo may be called
concurrently, so we need to create a new Hasher on
each hash call. Thanks @twmb for the find!

Epic: None

Release note: None

asg0451 committed:
changefeedccl: fix kafka sink v2 flush frequency

Previously, the flush frequency for the Kafka sink
v2 was hardcoded to 1 second. This correctly sets
it to the value specifed in the sink config.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128048)
<!-- Reviewable:end -->

asg0451 committed:
release-23.2: changefeedccl: add kafka sink v2

Changes required for backport. See the other PRs linked in this PR.

Epic: none

Release note: None

rharding6373 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for putting this together!
